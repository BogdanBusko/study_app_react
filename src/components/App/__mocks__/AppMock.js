export const userInfo = {
  data: {
    data: {
      user: {
        id: 1,
        type: 'user',
        attributes: {
          first_name: 'John',
          last_name: 'Doe',
          full_name: 'John Doe'
        }
      }
    }
  }
}
