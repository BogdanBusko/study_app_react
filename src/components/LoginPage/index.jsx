import { useForm } from 'react-hook-form';
import { useDispatch, useSelector } from 'react-redux';
import {
  Grid,
  TextField,
  Button
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

import createSession from '../../redux/actions/loginUser';

const useStyles = makeStyles({
  root: {
    paddingTop: '20px',
  },
})

const LoginPage = () => {
  const { register, handleSubmit, errors } = useForm();
  const wrongPassOrEmail = useSelector((state) => state.auth.wrongPassOrEmail);
  const dispatch = useDispatch();
  const classes = useStyles();

  const onSubmit = (data) => {
    dispatch(createSession(data));
  }

  return(
    <Grid
      container
      alignItems="center"
      justify="center"
      className={classes.root}
    >
      <form onSubmit={handleSubmit(onSubmit)}>
        <Grid container direction="column" spacing={2}>
          { wrongPassOrEmail &&
            <div className="error-label">
              Wrong password or email
            </div>
          }
          <TextField
            name="email"
            type="email"
            inputRef={register({required: "Email can't be blank"})}
            variant="outlined"
            label="Email"
            margin="normal"
          />
          { errors.email &&
            <label className="error-label">
              {errors.email.message}
            </label>
          }
          <TextField
            name="password"
            type="password"
            inputRef={register({required: "Password can't be blank"})}
            variant="outlined"
            label="Password"
            margin="normal"
          />
          { errors.password &&
            <label className="error-label">
              {errors.password.message}
            </label>
          }
          <Button variant="contained" type="submit">
            Login
          </Button>
        </Grid>
      </form>
    </Grid>
  );
}

export default LoginPage;
