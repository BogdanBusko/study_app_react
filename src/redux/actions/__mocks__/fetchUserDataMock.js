export const userInfo = {
  data: {
    data: {
      id: 1,
      type: 'user',
      attributes: {
        first_name: 'John',
        last_name: 'Doe',
        full_name: 'John Doe'
      }
    }
  }
}
