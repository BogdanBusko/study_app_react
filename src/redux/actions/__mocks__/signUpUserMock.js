export const registrationInfo = {
  data: {
    data: {
      id: '1',
      type: 'registration',
      attributes: {
        token: 'eyJhbGciOiJIUzI1NiJ9.eyJkYXRhIjp7InVzZXJfaWQiOjF9fQ.2MsjQvcIt3wDxROaQN21aMFNgtdYZCYCpJrlNwmYkcc',
        user: {
          data: {
            id: 1,
            type: 'user',
            attributes: {
              first_name: 'John',
              last_name: 'Doe',
              full_name: 'John Doe'
            }
          }
        }
      }
    }
  }
}
