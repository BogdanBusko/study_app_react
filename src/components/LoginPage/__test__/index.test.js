import React from 'react';
import { fireEvent, waitFor, act } from '@testing-library/react';

import fetchData from '../../../config/axios';
import { customRender } from '../../../test_helpers/helpers';
import LoginPage from '../index';
import * as mock from '../__mocks__/LoginPageMock';

jest.mock('../../../config/axios');

it("renders correctly", () => {
  const { asFragment } = customRender(<LoginPage/>);

  expect(asFragment(<LoginPage />)).toMatchSnapshot();
});

describe("with invalid data", () => {
  it("does not submit form when user does not enter email and password", async () => {
    const { queryByText, getByTestId } = customRender(<LoginPage/>);

    fireEvent.click(getByTestId('login-button'));

    await waitFor(() => expect(queryByText(/Email can't be blank/i)).toBeTruthy());
    await waitFor(() => expect(queryByText(/Password can't be blank/i)).toBeTruthy());
  });

  it("displays message 'Wrong password or email' if user enter wrong password or email and reset password field", async () => {
    const { queryByText, getByTestId } = customRender(<LoginPage/>);
    const emailInput = getByTestId('email').querySelector('input');
    const passwordInput = getByTestId('password').querySelector('input');

    fetchData.post.mockImplementationOnce(() => Promise.reject(new Error()));

    fireEvent.change(emailInput, { target: { value: 'email@email.com' } });
    fireEvent.change(passwordInput, { target: { value: '123456' } });

    await act(async () => fireEvent.click(getByTestId('login-button')));

    await waitFor(() => expect(queryByText(/Wrong password or email/i)).toBeTruthy());
    await waitFor(() => expect(passwordInput.value).toBeFalsy());
  });
})

describe("with valid data", () => {
  it("seves user token to localStorage", async () => {
    const { getByTestId } = customRender(<LoginPage />);
    const emailInput = getByTestId('email').querySelector('input');
    const passwordInput = getByTestId('password').querySelector('input');

    fetchData.post.mockResolvedValue(mock.successLogin);

    fireEvent.change(emailInput, { target: { value: 'email@email.com' } });
    fireEvent.change(passwordInput, { target: { value: '123456' } });

    await act(async () => fireEvent.click(getByTestId('login-button')));

    await waitFor(() =>
      expect(localStorage.getItem('token')).toEqual(mock.successLogin.data.data.attributes.token)
    );
  });
});
