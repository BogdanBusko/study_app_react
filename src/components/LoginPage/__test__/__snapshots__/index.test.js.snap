// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-root-1 MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <form>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          data-testid="email"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Email
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="email"
              type="email"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-4"
              >
                <span>
                  Email
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          data-testid="password"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-4"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained"
          data-testid="login-button"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Login
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
